// Code generated by ent, DO NOT EDIT.

package ent

import (
	"assignment4/ent/contactinfo"
	"assignment4/ent/predicate"
	"context"

	"entgo.io/ent/dialect/sql"
	"entgo.io/ent/dialect/sql/sqlgraph"
	"entgo.io/ent/schema/field"
)

// ContactInfoDelete is the builder for deleting a ContactInfo entity.
type ContactInfoDelete struct {
	config
	hooks    []Hook
	mutation *ContactInfoMutation
}

// Where appends a list predicates to the ContactInfoDelete builder.
func (cid *ContactInfoDelete) Where(ps ...predicate.ContactInfo) *ContactInfoDelete {
	cid.mutation.Where(ps...)
	return cid
}

// Exec executes the deletion query and returns how many vertices were deleted.
func (cid *ContactInfoDelete) Exec(ctx context.Context) (int, error) {
	return withHooks(ctx, cid.sqlExec, cid.mutation, cid.hooks)
}

// ExecX is like Exec, but panics if an error occurs.
func (cid *ContactInfoDelete) ExecX(ctx context.Context) int {
	n, err := cid.Exec(ctx)
	if err != nil {
		panic(err)
	}
	return n
}

func (cid *ContactInfoDelete) sqlExec(ctx context.Context) (int, error) {
	_spec := sqlgraph.NewDeleteSpec(contactinfo.Table, sqlgraph.NewFieldSpec(contactinfo.FieldID, field.TypeInt))
	if ps := cid.mutation.predicates; len(ps) > 0 {
		_spec.Predicate = func(selector *sql.Selector) {
			for i := range ps {
				ps[i](selector)
			}
		}
	}
	affected, err := sqlgraph.DeleteNodes(ctx, cid.driver, _spec)
	if err != nil && sqlgraph.IsConstraintError(err) {
		err = &ConstraintError{msg: err.Error(), wrap: err}
	}
	cid.mutation.done = true
	return affected, err
}

// ContactInfoDeleteOne is the builder for deleting a single ContactInfo entity.
type ContactInfoDeleteOne struct {
	cid *ContactInfoDelete
}

// Where appends a list predicates to the ContactInfoDelete builder.
func (cido *ContactInfoDeleteOne) Where(ps ...predicate.ContactInfo) *ContactInfoDeleteOne {
	cido.cid.mutation.Where(ps...)
	return cido
}

// Exec executes the deletion query.
func (cido *ContactInfoDeleteOne) Exec(ctx context.Context) error {
	n, err := cido.cid.Exec(ctx)
	switch {
	case err != nil:
		return err
	case n == 0:
		return &NotFoundError{contactinfo.Label}
	default:
		return nil
	}
}

// ExecX is like Exec, but panics if an error occurs.
func (cido *ContactInfoDeleteOne) ExecX(ctx context.Context) {
	if err := cido.Exec(ctx); err != nil {
		panic(err)
	}
}
